tensorflow/python/ops/nn_impl.py

```diff
@@ -550,6 +550,48 @@ def grad(dy):
   return features * math_ops.sigmoid(features), grad
 
 
+@tf_export("nn.tanhexp")
+@dispatch.add_dispatch_support
+@custom_gradient.custom_gradient
+def tanhexp(features):
+
+  """Computes the TanhExp activation function: `x * tanh(exp(x))`.
+
+  The TanhExp activation function was introduced in 
+  "TanhExp: A Smooth Activation Function with High Convergence
+  Speed for Lightweight Neural Networks"
+  [Xinyu et al. 2020](https://arxiv.org/pdf/2003.09855.pdf)
+
+  Args:
+    features: A `Tensor` representing preactivation values.
+    name: A name for the operation (optional).
+
+  Returns:
+    The activation value.
+  """
+
+  if isinstance(features, int):
+    features = ops.convert_to_tensor(features, name="features", dtype ='double')
+  else:
+    features = ops.convert_to_tensor(features, name="features")
+
+  def grad(dy):
+    """Gradient for the TanhExp activation function"""
+    # Naively, x * tf.nn.tanh(exp(x)) requires keeping both x and 
+    # tanh(exp(x)) around for backprop, effectively doubling the tensor's 
+    # memory consumption. We use a control dependency here so that 
+    # tanh(exp(x)) is re-computed during backprop (the control dep prevents 
+    # it being de-duped with the forward pass) and we can free the 
```

Comment:
Why not `tf.recompute_grad`? That will save a bit more memory executing eagerly and avoids repeating the gradient definitions.

Please also add doctested examples to the tf.nn symbol and a unit test. If you're defining a custom gradient it'd need numeric gradient checking too (probably fine to omit for recompute_grad).